File: app/backend/src/domain/aggregates/order/order.rs
use super::{CustomerInfo, OrderItem, OrderPricing, PaymentInfo, ShippingInfo};
use crate::domain::entities::DeliveryInfo;
use crate::domain::error::DomainError;
use crate::domain::value_objects::*;
use chrono::{DateTime, Utc};

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum OrderStatus {
    Pending,    // Order created, awaiting payment
    Paid,       // Payment confirmed
    Processing, // Order being prepared
    Shipped,    // Order dispatched
    Delivered,  // Order delivered
    Cancelled,  // Order cancelled
    Refunded,   // Order refunded
}

impl std::fmt::Display for OrderStatus {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            OrderStatus::Pending => write!(f, "pending"),
            OrderStatus::Paid => write!(f, "paid"),
            OrderStatus::Processing => write!(f, "processing"),
            OrderStatus::Shipped => write!(f, "shipped"),
            OrderStatus::Delivered => write!(f, "delivered"),
            OrderStatus::Cancelled => write!(f, "cancelled"),
            OrderStatus::Refunded => write!(f, "refunded"),
        }
    }
}

impl std::str::FromStr for OrderStatus {
    type Err = crate::domain::error::DomainError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "pending" => Ok(OrderStatus::Pending),
            "paid" => Ok(OrderStatus::Paid),
            "processing" => Ok(OrderStatus::Processing),
            "shipped" => Ok(OrderStatus::Shipped),
            "delivered" => Ok(OrderStatus::Delivered),
            "cancelled" => Ok(OrderStatus::Cancelled),
            "refunded" => Ok(OrderStatus::Refunded),
            _ => Err(crate::domain::error::DomainError::InvalidProductData(
                format!("Invalid order status: {}", s),
            )),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct OrderTimestamps {
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub paid_at: Option<DateTime<Utc>>,
    pub shipped_at: Option<DateTime<Utc>>,
    pub delivered_at: Option<DateTime<Utc>>,
    pub cancelled_at: Option<DateTime<Utc>>,
}

#[derive(Debug, Clone)]
pub struct Order {
    pub id: OrderId,
    pub order_number: OrderNumber,
    pub customer_info: CustomerInfo,
    pub items: Vec<OrderItem>,
    pub shipping_info: ShippingInfo,
    pub payment_info: PaymentInfo,
    pub pricing: OrderPricing,
    pub status: OrderStatus,
    pub timestamps: OrderTimestamps,
    pub delivery_info: Option<DeliveryInfo>,
    pub notes: Option<String>,
}

impl Order {
    pub fn new(
        order_number: OrderNumber,
        customer_info: CustomerInfo,
        items: Vec<OrderItem>,
        shipping_info: ShippingInfo,
        payment_info: PaymentInfo,
    ) -> Result<Self, DomainError> {
        if items.is_empty() {
            return Err(DomainError::InvalidProductData(
                "Order must have at least one item".to_string(),
            ));
        }

        let pricing = Self::calculate_pricing(&items, &shipping_info, &payment_info)?;
        let now = Utc::now();

        Ok(Order {
            id: OrderId::new(),
            order_number,
            customer_info,
            items,
            shipping_info,
            payment_info,
            pricing,
            status: OrderStatus::Pending,
            timestamps: OrderTimestamps {
                created_at: now,
                updated_at: now,
                paid_at: None,
                shipped_at: None,
                delivered_at: None,
                cancelled_at: None,
            },
            delivery_info: None,
            notes: None,
        })
    }

    pub fn update_status(&mut self, new_status: OrderStatus) -> Result<(), DomainError> {
        self.validate_status_transition(&new_status)?;

        let now = Utc::now();
        self.status = new_status.clone();
        self.timestamps.updated_at = now;

        // Update relevant timestamps
        match new_status {
            OrderStatus::Paid => self.timestamps.paid_at = Some(now),
            OrderStatus::Shipped => self.timestamps.shipped_at = Some(now),
            OrderStatus::Delivered => self.timestamps.delivered_at = Some(now),
            OrderStatus::Cancelled => self.timestamps.cancelled_at = Some(now),
            _ => {}
        }

        Ok(())
    }

    pub fn add_delivery_info(&mut self, delivery_info: DeliveryInfo) -> Result<(), DomainError> {
        if !matches!(self.status, OrderStatus::Paid | OrderStatus::Processing) {
            return Err(DomainError::InvalidProductData(format!(
                "Cannot add delivery info to order in status {:?}",
                self.status
            )));
        }

        self.delivery_info = Some(delivery_info);
        self.timestamps.updated_at = Utc::now();
        Ok(())
    }

    pub fn cancel(&mut self, reason: String) -> Result<(), DomainError> {
        if matches!(
            self.status,
            OrderStatus::Delivered | OrderStatus::Cancelled | OrderStatus::Refunded
        ) {
            return Err(DomainError::InvalidProductData(format!(
                "Cannot cancel order in status {:?}",
                self.status
            )));
        }

        self.status = OrderStatus::Cancelled;
        self.notes = Some(reason);
        self.timestamps.cancelled_at = Some(Utc::now());
        self.timestamps.updated_at = Utc::now();

        Ok(())
    }

    pub fn add_note(&mut self, note: String) -> Result<(), DomainError> {
        if note.len() > 1000 {
            return Err(DomainError::InvalidProductData(
                "Note cannot exceed 1000 characters".to_string(),
            ));
        }

        self.notes = Some(note);
        self.timestamps.updated_at = Utc::now();
        Ok(())
    }

    fn validate_status_transition(&self, new_status: &OrderStatus) -> Result<(), DomainError> {
        use OrderStatus::*;

        let valid = match (&self.status, new_status) {
            (Pending, Paid) => true,
            (Pending, Cancelled) => true,
            (Paid, Processing) => true,
            (Paid, Cancelled) => true,
            (Processing, Shipped) => true,
            (Processing, Cancelled) => true,
            (Shipped, Delivered) => true,
            (_, Refunded) => matches!(self.status, Paid | Processing | Shipped | Delivered),
            _ => false,
        };

        if !valid {
            return Err(DomainError::InvalidProductData(format!(
                "Invalid status transition from {:?} to {:?}",
                self.status, new_status
            )));
        }

        Ok(())
    }

    fn calculate_pricing(
        items: &[OrderItem],
        shipping_info: &ShippingInfo,
        payment_info: &PaymentInfo,
    ) -> Result<OrderPricing, DomainError> {
        let mut subtotal = Money::zero();
        for item in items {
            subtotal = subtotal.add(item.subtotal()?)?;
        }

        let total_before_tax = subtotal.add(shipping_info.fee)?.add(payment_info.fee)?;
        let tax_amount = total_before_tax.tax_amount();
        let total = total_before_tax.add(tax_amount)?;

        Ok(OrderPricing {
            subtotal,
            shipping_fee: shipping_info.fee,
            payment_fee: payment_info.fee,
            tax_amount,
            total,
        })
    }

    pub fn can_be_cancelled(&self) -> bool {
        !matches!(
            self.status,
            OrderStatus::Delivered | OrderStatus::Cancelled | OrderStatus::Refunded
        )
    }

    pub fn can_be_modified(&self) -> bool {
        matches!(self.status, OrderStatus::Pending)
    }

    pub fn total_item_count(&self) -> i32 {
        self.items.iter().map(|item| item.quantity).sum()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::domain::value_objects::*;

    fn create_test_order() -> Order {
        let order_number = OrderNumber::generate(2024, 1);
        let customer_info = CustomerInfo::new(
            PersonalInfo::new(
                FirstName::new("太郎".to_string()).unwrap(),
                LastName::new("田中".to_string()).unwrap(),
            ),
            Email::new("test@example.com".to_string()).unwrap(),
            PhoneNumber::new("090-1234-5678".to_string()).unwrap(),
        );

        let items = vec![
            OrderItem::new(
                SKUId::new(),
                SKUCode::new("TEST-001".to_string()).unwrap(),
                ProductName::new("Test Product".to_string()).unwrap(),
                SKUName::new("Test SKU".to_string()).unwrap(),
                Money::from_yen(1000),
                2,
            )
            .unwrap(),
        ];

        let shipping_info = ShippingInfo::new(
            ShippingMethodId::new("standard".to_string()).unwrap(),
            "Standard Shipping".to_string(),
            Money::from_yen(500),
            Address::new(
                "123-4567".to_string(),
                "Tokyo".to_string(),
                "Shibuya".to_string(),
                "1-2-3".to_string(),
                None,
            )
            .unwrap(),
        );

        let payment_info = PaymentInfo::new(
            PaymentMethodId::new(),
            "Credit Card".to_string(),
            Money::from_yen(100),
            None,
        );

        Order::new(
            order_number,
            customer_info,
            items,
            shipping_info,
            payment_info,
        )
        .unwrap()
    }

    #[test]
    fn test_create_order() {
        let order = create_test_order();
        assert_eq!(order.status, OrderStatus::Pending);
        assert_eq!(order.items.len(), 1);
        assert!(order.delivery_info.is_none());
    }

    #[test]
    fn test_status_transitions() {
        let mut order = create_test_order();

        // Pending -> Paid
        assert!(order.update_status(OrderStatus::Paid).is_ok());
        assert_eq!(order.status, OrderStatus::Paid);
        assert!(order.timestamps.paid_at.is_some());

        // Paid -> Processing
        assert!(order.update_status(OrderStatus::Processing).is_ok());
        assert_eq!(order.status, OrderStatus::Processing);

        // Processing -> Shipped
        assert!(order.update_status(OrderStatus::Shipped).is_ok());
        assert_eq!(order.status, OrderStatus::Shipped);
        assert!(order.timestamps.shipped_at.is_some());

        // Shipped -> Delivered
        assert!(order.update_status(OrderStatus::Delivered).is_ok());
        assert_eq!(order.status, OrderStatus::Delivered);
        assert!(order.timestamps.delivered_at.is_some());
    }

    #[test]
    fn test_invalid_status_transitions() {
        let mut order = create_test_order();

        // Cannot go from Pending to Delivered
        assert!(order.update_status(OrderStatus::Delivered).is_err());

        // Cannot go from Pending to Shipped
        assert!(order.update_status(OrderStatus::Shipped).is_err());
    }

    #[test]
    fn test_cancel_order() {
        let mut order = create_test_order();

        assert!(order.cancel("Customer request".to_string()).is_ok());
        assert_eq!(order.status, OrderStatus::Cancelled);
        assert!(order.timestamps.cancelled_at.is_some());
        assert_eq!(order.notes, Some("Customer request".to_string()));
    }

    #[test]
    fn test_cannot_cancel_delivered_order() {
        let mut order = create_test_order();
        order.status = OrderStatus::Delivered;

        assert!(order.cancel("Too late".to_string()).is_err());
    }

    #[test]
    fn test_empty_items_error() {
        let order_number = OrderNumber::generate(2024, 1);
        let customer_info = CustomerInfo::new(
            PersonalInfo::new(
                FirstName::new("太郎".to_string()).unwrap(),
                LastName::new("田中".to_string()).unwrap(),
            ),
            Email::new("test@example.com".to_string()).unwrap(),
            PhoneNumber::new("090-1234-5678".to_string()).unwrap(),
        );

        let items = vec![];

        let shipping_info = ShippingInfo::new(
            ShippingMethodId::new("standard".to_string()).unwrap(),
            "Standard Shipping".to_string(),
            Money::from_yen(500),
            Address::new(
                "123-4567".to_string(),
                "Tokyo".to_string(),
                "Shibuya".to_string(),
                "1-2-3".to_string(),
                None,
            )
            .unwrap(),
        );

        let payment_info = PaymentInfo::new(
            PaymentMethodId::new(),
            "Credit Card".to_string(),
            Money::from_yen(100),
            None,
        );

        let result = Order::new(
            order_number,
            customer_info,
            items,
            shipping_info,
            payment_info,
        );
        assert!(result.is_err());
    }
}

File: app/backend-ts/src/domain/aggregates/order/order.ts
import { SKU } from '$domain/entities';
import { DomainError } from '$domain/errors/domain.error';
import { Address, Money, OrderId, OrderNumber } from '$domain/value-objects';

import { CustomerInfo } from './customer-info';
import { OrderItem } from './order-item';
import { OrderPricing } from './order-pricing';
import { PaymentInfo } from './payment-info';
import { ShippingInfo } from './shipping-info';

export enum OrderStatus {
  Pending = 'pending',
  Paid = 'paid',
  Processing = 'processing',
  Shipped = 'shipped',
  Delivered = 'delivered',
  Cancelled = 'cancelled',
  Refunded = 'refunded',
}

export class OrderTimestamps {
  constructor(
    public readonly createdAt: Date,
    public updatedAt: Date,
    public paidAt?: Date,
    public shippedAt?: Date,
    public deliveredAt?: Date,
    public cancelledAt?: Date,
  ) {}

  update(): void {
    this.updatedAt = new Date();
  }

  markPaid(): void {
    this.paidAt = new Date();
    this.update();
  }

  markShipped(): void {
    this.shippedAt = new Date();
    this.update();
  }

  markDelivered(): void {
    this.deliveredAt = new Date();
    this.update();
  }

  markCancelled(): void {
    this.cancelledAt = new Date();
    this.update();
  }
}

export class Order {
  private constructor(
    private readonly id: OrderId,
    private readonly orderNumber: OrderNumber,
    private readonly customerInfo: CustomerInfo,
    private readonly items: OrderItem[],
    private readonly shippingInfo: ShippingInfo,
    private readonly paymentInfo: PaymentInfo,
    private readonly pricing: OrderPricing,
    private status: OrderStatus,
    private readonly timestamps: OrderTimestamps,
    private notes?: string,
  ) {}

  static create(
    id: OrderId,
    orderNumber: OrderNumber,
    customerInfo: CustomerInfo,
    items: OrderItem[],
    shippingInfo: ShippingInfo,
    paymentInfo: PaymentInfo,
  ): Order {
    if (items.length === 0) {
      throw new DomainError('Order must have at least one item');
    }

    // Calculate pricing
    const subtotal = items
      .map((item) => item.subtotal())
      .reduce((total, itemSubtotal) => total.add(itemSubtotal));

    const pricing = OrderPricing.calculate(
      subtotal,
      shippingInfo.getShippingFee(),
      paymentInfo.getPaymentFee(),
    );

    const timestamps = new OrderTimestamps(new Date(), new Date());

    return new Order(
      id,
      orderNumber,
      customerInfo,
      items,
      shippingInfo,
      paymentInfo,
      pricing,
      OrderStatus.Pending,
      timestamps,
    );
  }

  static createFromCommand(
    orderId: OrderId,
    orderNumber: OrderNumber,
    customerInfo: CustomerInfo,
    commandItems: Array<{ skuId: string; quantity: number }>,
    availableSkus: SKU[], // 既存のSKUエンティティを使用
    shippingMethodData: {
      id: string;
      name: string;
      fee: Money;
      isActive: boolean;
    },
    shippingAddress: Address,
    paymentMethodData: {
      id: string;
      name: string;
      fee: Money;
      isActive: boolean;
    },
  ): Order {
    // 既存のSKUエンティティのビジネスロジックを活用
    const orderItems = this.validateAndCreateOrderItems(
      commandItems,
      availableSkus,
    );

    // 拡張された既存クラスを使用
    const shippingInfo = ShippingInfo.createFromMethod(
      shippingMethodData,
      shippingAddress,
    );
    const paymentInfo = PaymentInfo.createFromMethod(paymentMethodData);

    return Order.create(
      orderId,
      orderNumber,
      customerInfo,
      orderItems,
      shippingInfo,
      paymentInfo,
    );
  }

  private static validateAndCreateOrderItems(
    commandItems: Array<{ skuId: string; quantity: number }>,
    availableSkus: SKU[],
  ): OrderItem[] {
    const orderItems: OrderItem[] = [];

    for (const commandItem of commandItems) {
      const sku = availableSkus.find(
        (s) => s.getId().value() === commandItem.skuId,
      );

      if (!sku) {
        throw new DomainError(`SKU not found: ${commandItem.skuId}`);
      }

      // 既存のSKUエンティティのビジネスロジックを活用
      if (!sku.isPurchasable()) {
        throw new DomainError(
          `SKU ${commandItem.skuId} is not available for purchase`,
        );
      }

      if (sku.availableQuantity() < commandItem.quantity) {
        throw new DomainError(
          `Insufficient stock for SKU ${commandItem.skuId}. ` +
            `Requested: ${commandItem.quantity}, Available: ${sku.availableQuantity()}`,
        );
      }

      const orderItem = OrderItem.create(
        sku.getId(),
        sku.getProductId(),
        sku.getName(),
        sku.getName(),
        sku.currentPrice(),
        commandItem.quantity,
      );

      orderItems.push(orderItem);
    }

    return orderItems;
  }

  // Status management
  updateStatus(newStatus: OrderStatus): void {
    this.validateStatusTransition(newStatus);

    // const oldStatus = this.status; // Reserved for future use (e.g., logging)
    this.status = newStatus;

    // Update timestamps based on status
    switch (newStatus) {
      case OrderStatus.Paid:
        this.timestamps.markPaid();
        break;
      case OrderStatus.Shipped:
        this.timestamps.markShipped();
        break;
      case OrderStatus.Delivered:
        this.timestamps.markDelivered();
        break;
      case OrderStatus.Cancelled:
        this.timestamps.markCancelled();
        break;
      default:
        this.timestamps.update();
        break;
    }
  }

  cancel(reason?: string): void {
    if (!this.canBeCancelled()) {
      throw new DomainError(`Cannot cancel order with status: ${this.status}`);
    }

    this.updateStatus(OrderStatus.Cancelled);

    if (reason && reason.trim().length > 0) {
      this.addNote(`Cancelled: ${reason.trim()}`);
    }
  }

  // Content management
  addNote(note: string): void {
    if (!note || note.trim().length === 0) {
      throw new DomainError('Note cannot be empty');
    }
    if (note.trim().length > 1000) {
      throw new DomainError('Note cannot exceed 1000 characters');
    }

    const trimmedNote = note.trim();
    if (this.notes) {
      this.notes += `\n${trimmedNote}`;
    } else {
      this.notes = trimmedNote;
    }

    this.timestamps.update();
  }

  // Business logic
  canBeCancelled(): boolean {
    return (
      this.status !== OrderStatus.Delivered &&
      this.status !== OrderStatus.Cancelled &&
      this.status !== OrderStatus.Refunded
    );
  }

  canBeModified(): boolean {
    return this.status === OrderStatus.Pending;
  }

  totalItemCount(): number {
    return this.items.reduce((total, item) => total + item.getQuantity(), 0);
  }

  private validateStatusTransition(newStatus: OrderStatus): void {
    const validTransitions: Record<OrderStatus, OrderStatus[]> = {
      [OrderStatus.Pending]: [OrderStatus.Paid, OrderStatus.Cancelled],
      [OrderStatus.Paid]: [OrderStatus.Processing, OrderStatus.Cancelled],
      [OrderStatus.Processing]: [OrderStatus.Shipped, OrderStatus.Cancelled],
      [OrderStatus.Shipped]: [OrderStatus.Delivered],
      [OrderStatus.Delivered]: [], // No transitions from delivered
      [OrderStatus.Cancelled]: [OrderStatus.Refunded],
      [OrderStatus.Refunded]: [], // No transitions from refunded
    };

    const allowedStatuses = validTransitions[this.status];
    if (!allowedStatuses.includes(newStatus)) {
      throw new DomainError(
        `Invalid status transition from ${this.status} to ${newStatus}`,
      );
    }
  }

  // Getters
  getId(): OrderId {
    return this.id;
  }

  getOrderNumber(): OrderNumber {
    return this.orderNumber;
  }

  getCustomerInfo(): CustomerInfo {
    return this.customerInfo;
  }

  getItems(): OrderItem[] {
    return [...this.items];
  }

  getShippingInfo(): ShippingInfo {
    return this.shippingInfo;
  }

  getPaymentInfo(): PaymentInfo {
    return this.paymentInfo;
  }

  getPricing(): OrderPricing {
    return this.pricing;
  }

  getStatus(): OrderStatus {
    return this.status;
  }

  getTimestamps(): OrderTimestamps {
    return this.timestamps;
  }

  getNotes(): string | undefined {
    return this.notes;
  }
}
